### PR TITLE
Add CPPO benchmarks and trainer

### LINKED ISSUE #28: CPPO

### PR DESCRIPTION
**Pull request by PythonSemicolon**
Closes #28.

_Note: running tests locally, will update with comments once complete._

### CODE REVIEW

File: benchmarks/cppo/cppo.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    # check if the reward models are present either in the path or in the hub
+    if training_args.reward_model_path is not None:
+        for i in range(len(continual_dataset)):
+            reward_path = training_args.reward_model_path + '_' + str(i)
+            # first check the hub if the model is present
+            try:
+                AutoModelForSequenceClassification.from_pretrained(
+                    reward_path, num_labels=1
+                )
+            except:
```
[the remark this answers is not in the document]
Comment (Jacob-Chmura) [reply benchmarks/cppo/cppo.py:109]: Leaving this for now as it is present in all our other benchmark code. Will be easier to unify

**Comment by Jacob-Chmura** [comment]
[diff.txt](https://github.com/user-attachments/files/19979072/diff.txt)

File: benchmarks/cppo/cppo_trainer.py [hunk 2]
```diff
@@ -0,0 +1,1330 @@
...
+    logp_std = old_logprobs.std(dim=-1)
+    rewards_mean = old_rewards.mean(dim=-1)
+    rewards_std = old_rewards.std(dim=-1)
+
+    N = old_logprobs.shape[0]
+    device = old_logprobs.device
+
+    coef_learn = torch.ones_like(old_logprobs[:, 0])
+    coef_reg = torch.ones_like(old_logprobs[:, 0])
+    mask = mask[0]  # its a list of a tensor
```
Comment (Jacob-Chmura) [inline benchmarks/cppo/cppo_trainer.py:1297]: I'm pretty sure we need to `torch.cat(mask, dim=0)` in case we have multiple batches as in: https://github.com/ComplexData-MILA/AIF-Gen/blob/3398a0a9764d7bef50e67e667c3bb90fe539bba8/benchmarks/cppo/cppo_trainer.py#L681-L687

And I think we can do the cat up there on line 681 so that this helper function just gets a tensor for the mask

File: benchmarks/cppo/cppo_trainer.py [hunk 3]
```diff
@@ -0,0 +1,1330 @@
...
+    threshold12 = (rewards_mean - threshold * rewards_std).to(device)
+    threshold21 = (logp_mean + threshold * logp_std).to(device)
+    threshold22 = (logp_mean - threshold * logp_std).to(device)
+
+    cond11 = (
+        torch.gather(old_rewards, dim=-1, index=length.unsqueeze(-1)) > threshold11
+    )[0]
+    cond12 = (
+        torch.gather(old_rewards, dim=-1, index=length.unsqueeze(-1)) < threshold12
+    )[0]
```
Comment (Jacob-Chmura) [inline benchmarks/cppo/cppo_trainer.py:1310]: Would the `[0]` still be needed assuming we did the mask `torch.cat` as above?

File: benchmarks/cppo/cppo_trainer.py [hunk 4]
```diff
@@ -0,0 +1,1330 @@
...
+                            mask = ~padding_mask[micro_batch_inds]
+                            vf_loss = 0.5 * masked_mean(vf_loss_max, mask)
+                            vf_clipfrac = masked_mean(
+                                (vf_losses2 > vf_losses1).float(), mask
+                            )
+
+                            def _get_mask(coef: Optional[Tensor]) -> Tensor:
+                                if coef is None:
+                                    return mask
+                                _mask = mask * coef[micro_batch_inds].unsqueeze(1)
```
Comment (Jacob-Chmura) [inline benchmarks/cppo/cppo_trainer.py:814]: Is the unsqueeze needed assuming the mask is `torch.cat` as I described?

Comment (EMZEDI) [reply benchmarks/cppo/cppo_trainer.py:814]: seems like it is still needed?

File: benchmarks/cppo/cppo_trainer.py (hunk 2 above)
Comment (EMZEDI) [reply benchmarks/cppo/cppo_trainer.py:1297]: Applied. I appreciate a second review.

**Review by Jacob-Chmura — APPROVED** [review]
